import_export can force fields to be non-negative

mnlevy1981 opened the pull request:
If the atm -> ice scalar map is not positive definite, then ensure Qa and rhoa
are non-negative by setting all negative values to 0. Similarly, if the atm ->
ice flux map is not positive definite, ensure swvdr, swidr, swvdf, swidf,
frain, and fsnow are all non-negative.

The implementation is a little kludgy: component models in CESM do not have
access to public data in the coupler (otherwise this introduces circular
dependencies), so we can't just query the driver to find out about the map.
Instead, I introduce two new namelist variables in &setup_nml and buildnml sets
them based on the ATM2OCN map names in env_run.xml. Currently, the logic is "if
'patc' shows up in the file name, it's a patch map and therefore not positive
definite. Otherwise it's a positive definite map."

Fixes #28 


****

### Testing

I ran `aux_cice` with the current `master` branch to generate baselines. For these baselines, four tests failed:

```
4	FAIL
	1	COMPARE (FAIL)
      1 ERP.f19_g17_rx1.DTEST.cheyenne_intel.cice-default
	3	CREATE_NEWCASE (FAIL)
      1 ERS_Ld5.f19_g17.ETEST.cheyenne_intel.cice-default
      1 ERS_Ld5.T62_t061.DTEST.cheyenne_intel.cice-default
      1 ERS_Ld7.f19_g17.B1850.cheyenne_intel.cice-pop
```

My sandbox did not have CAM available, so that explains the `ETEST` and `B1850` failures. The `T62_t061` test failed with `ERROR: No variable grid found in case` in the `TestStatus.log`. The `ERP` test is showing the following in `cice.h.0001-01-11.nc.base.cprnc.out`

```
SUMMARY of cprnc:
 A total number of    116 fields were compared
          of which      0 had non-zero differences
               and     14 had differences in fill patterns
               and      0 had different dimension sizes
 A total number of      0 fields could not be analyzed
 A total number of      0 fields on file 1 were not found on file2.
  diff_test: the two files seem to be DIFFERENT
```

and similarly in `cpl.hi.0001-01-12-00000.nc.base.cprnc.out`

```
SUMMARY of cprnc:
 A total number of    319 fields were compared
          of which      0 had non-zero differences
               and     49 had differences in fill patterns
               and      0 had different dimension sizes
 A total number of      0 fields could not be analyzed
 A total number of      0 fields on file 1 were not found on file2.
  diff_test: the two files seem to be DIFFERENT
```

Again, that was just from generating a baseline. When comparing my branch, I get the same four failures (for the same reason). Nine tests run to completion, and all of them fail `NLCOMP` due to the addition of two new namelist variables, but the important comparison is

```
113	PASS
	9	BASELINE (PASS)
      1 ERI.f19_g17_rx1.G.cheyenne_intel.pop-cice
      1 ERI.T62_g17.DTEST.cheyenne_intel.cice-default
      1 ERP.f19_g17_rx1.DTEST.cheyenne_intel.cice-default
      1 ERS_D_Ld5.T62_g17.DTEST.cheyenne_intel.cice-default
      1 ERS.f19_g17_rx1.G1850ECO.cheyenne_intel.pop-cice
      1 ERS_Ld5.T62_s11.DTEST.cheyenne_intel.cice-default
      1 ERS_Lm3.T62_g37.G.cheyenne_intel.pop-cice
      1 PET.f19_g17_rx1.G.cheyenne_intel.pop-cice
      1 PET_Ld2.T62_g17.DTEST.cheyenne_intel.cice-default
```

Namely, the nine tests that run to completion are bit-for-bit identical with the previous CICE `master` branch.

I also verified that a `SMS_Ld1.TL319_t13.GIAF_JRA_HR.cheyenne_intel.pop-performance_eval` failed on the current `master` with `ERROR: ice: Vertical thermo error` but successfully complete with this branch.

****

### Final note

As mentioned above, this is a somewhat kludgy implementation. I'm happy to update this PR with a cleaner implementation if someone can come up with a better way of doing things... but given the limitation that CICE can't access the mapping data in the driver once the model is running, this seemed like the only option.

mnlevy1981 committed:
import_export can force fields to be non-negative

If the atm -> ice scalar map is not positive definite, then ensure Qa and rhoa
are non-negative by setting all negative values to 0. Similarly, if the atm ->
ice flux map is not positive definite, ensure swvdr, swidr, swvdf, swidf,
frain, and fsnow are all non-negative.

The implementation is a little kludgy: component models in CESM do not have
access to public data in the coupler (otherwise this introduces circular
dependencies), so we can't just query the driver to find out about the map.
Instead, I introduce two new namelist variables in &setup_nml and buildnml sets
them based on the ATM2OCN map names in env_run.xml. Currently, the logic is "if
'patc' shows up in the file name, it's a patch map and therefore not positive
definite. Otherwise it's a positive definite map."

dabail10 commented on cime_config/buildnml:
```diff
@@ -121,6 +121,16 @@ def _create_namelists(case, confdir, infile, nmlgen):
             nmlgen.set_value('ice_ic', value="default")
 
+    # patch maps are not positive definite, aavg and blin are
+    if case.get_value("ATM2OCN_FMAPNAME").find('patc') == -1:
```
Really? It returns -1 if true?

dabail10 replied:
Oh wait, the logic is the reverse of what I was thinking. I would rather have the default to always be .true. and only change it when 'patc' is found. Doesn't namelist_definition_cice.xml set these to .true. by default?

mnlevy1981 replied:
In `namelist_definition_cice.xml` I defined the variables with `skip_default_entry="true"` because I wanted to use python for the logic of "is this map a patch map or not?". So `buildnml` needs to set the variable to `.false.` for patch maps and `.true.` otherwise -- in python, `str.find(substr)` returns the [0-based] index of the first character matching `substr` or returns -1 if the `substr` is not found. So `case.get_value("ATM2OCN_FMAPNAME").find('patc') == -1` is true if the file name does not contain `patc`.

mnlevy1981 replied:
It's maybe a little confusing that the fortran variables track whether the map is positive [semi-]definite or not, given that the interesting case is when the map is NOT positive definite... but I didn't want to name the variables `atm2ice_smap_is_patch` because what happens if another non-positive definite map gets introduced? Either the variable name is misleading or it needs to be changed. On the other hand, `atm2ice_smap_is_not_pos_def` seems awfully clunky.

dabail10 commented:
This is probably fine. I just feel like the variables should always default to .true. and only be set to .false. in a few cases.

mnlevy1981 commented:
I've got an idea on how to do that, I'll update the PR in 15 min or so...

mnlevy1981 committed:
Drop skip_default_entry from atm2ice_ variables

Instead, default values for atm2ice_fmap_is_pos_def (and smap equivalent) are
both .true., but if fmaptype (or smaptype) is "patch" they are toggled to
false. The buildnml script sets config['fmaptype'] and config['smaptype'] based
on the str.find() command, though if other maps that are not positive definite
are introduced perhaps these if statements should be replaced with a function?

mnlevy1981 commented:
For testing this latest commit, I created a `G` compset and a `GIAF_JRA_HR` compset and verified that the two `atm2ice_` variables remained `.true.` in the former and `.false.` in the latter.

dabail10 reviewed (approved):
I like this better!
